src/dmd/target.d

```diff
@@ -1446,5 +1444,204 @@ struct TargetObjC
+            cppenv = TargetCPP.Runtime.Microsoft;
+    }
+    private extern(D):
+
+    void unknown(const(char)[] unk, const(char)* what)
+    {
+        import dmd.errors : error;
+        import dmd.root.string : toCStringThen;
+        import dmd.globals : Loc;
+        unk.toCStringThen!(p => error(Loc.initial,"unknown %s `%s` for `-target`", what, p.ptr));
+    }
+
+    void parseArch(const(char)[] arch)
+    {
+        bool matches(const(char)[] str)
+        {
+            import dmd.root.string : startsWith;
+            if (!arch.ptr.startsWith(str))
+                return false;
+            arch = arch[str.length-1 .. $-1];
+            return true;
+        }
+
+        if (matches("x86_64"))
+            is64bit = true;
+        else if (matches("x86"))
+            is64bit = false;
+        else if (matches("x64"))
+            is64bit = true;
+        else if (matches("x32"))
+        {
+            is64bit = true;
+            isLP64 = false;
+        }
+        else
+            return unknown(arch, "architecture");
+
+        if (!arch.length)
+            return;
+
+        switch (arch)
+        {
+            case "+sse2": cpu = CPU.sse2; break;
+            case "+avx":  cpu = CPU.avx;  break;
+            case "+avx2": cpu = CPU.avx2; break;
+            default:
+                unknown(arch, "architecture feature");
+        }
+    }
+
+    // try parsing vendor if present
+    bool tryParseVendor(const(char)[] vendor)
+    {
+        switch (vendor)
+        {
+            case "unknown": return true;
+            case "apple":   return true;
+            case "pc":      return true;
+            case "amd":     return true;
+            default:        return false;
+        }
+    }
+
+    Target.OS parseOS(const(char)[] _os, out ubyte _osMajor)
+    {
+        bool matches(const(char)[] str)
+        {
+            import dmd.root.string : startsWith;
+            if (!_os.ptr.startsWith(str))
+                return false;
+            _os = _os[str.length .. $];
+            return true;
+        }
+        if (_os == "freestanding")
+            return Target.OS.Freestanding;
+        Target.OS os;
+        _osMajor = 0;
+        if (matches("darwin"))
+            os = Target.OS.OSX;
+        else if (matches("dragonfly"))
+            os =  Target.OS.DragonFlyBSD;
+        else if (matches("freebsd"))
+            os =  Target.OS.FreeBSD;
+        else if (matches("openbsd"))
+            os =  Target.OS.OpenBSD;
+        else if (matches("linux"))
+            os =  Target.OS.linux;
+        else if (matches("windows"))
+            os =  Target.OS.Windows;
+        else
+        {
+            unknown(_os, "operating system");
+            return Target.OS.Freestanding;
+        }
+        while (_os.length)
+        {
+            if (!('0' < _os[0] && _os[0] < '9'))
+                break;
+            osMajor *= 10;
+            osMajor = cast(ubyte)((_os[0] - '0') + osMajor);
```

Comment:
I don't mean to spoil the party, but `core.checkedint`! :-)

Reply:
I'm not quite sure what you mean, but the cast is there because of integral promotions.

Reply:
osMajor can overflow multiple times with `*= 10`, for instance `darwin5000`

Reply:
if someone is silly enough to target `darwin5000` then they deserve what they get. Note that `osMajor` is only used for setting `FreeBSD#` where `#` is 10,11 or 12.